### PR TITLE
Implement scatterElements

### PR DESCRIPTION
**Pull request by BruceDai**
@huningxin @fdwr @shiyi9801 PTAL, thanks.

### CODE REVIEW

File: src/lib/validate-input.js [hunk 1]
```diff
@@ -611,6 +611,46 @@ export function validateGatherParams(input, indices, {axis = 0} = {}) {
...
+  if (!updates.shape.every((value, index) => value === indices.shape[index])) {
+    throw new Error(`Invalid updates value - updates shape should be same as indices shape.`);
+  }
+  if (axis !== undefined) {
+    if (!Number.isInteger(axis) || axis < 0) {
+      throw new Error(`The axis ${axis} should be an unsigned integer.`);
+    }
+    if (axis >= inputRank) {
+      throw new Error(`The axis ${axis} should be in the interval [0, ${inputRank}).`);
+    }
```
Comment (fdwr) [inline src/lib/validate-input.js:642]: Consider merging (like line 647).

```suggestion
    if (!Number.isInteger(axis) || axis < 0 || axis >= inputRank) {
      throw new Error(`The axis ${axis} should be an unsigned integer in the interval [0, ${inputRank}).`);
    }
```

File: src/scatter_elements.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+export function scatterElements(input, indices, updates, {axis = 0} = {}) {
+  validateScatterElementsParams(...arguments);
+  const output = identity(input);
+
+  for (let indicesIndex = 0; indicesIndex < sizeOfShape(indices.shape); ++indicesIndex) {
+    // output[indices[i, j, k, ...], j, k, ...] = updates[i, j, k, ...] // if axis == 0
+    // output[i, indices[i, j, k, ...], k, ...] = updates[i, j, k, ...] // if axis == 1
+    // output[i, j, indices[i, j, k, ...], ...] = updates[i, j, k, ...] // if axis == 2
+    const indicesLoc = indices.locationFromIndex(indicesIndex);
```
Comment (fdwr) [inline src/scatter_elements.js:24]:
```suggestion
    const indicesLocation = indices.locationFromIndex(indicesIndex);
```
Whole words please.

**Review by fdwr — APPROVED** [review]
Thanks Bruce. Some small comments, and then LGTM.

**Comment by BruceDai** [comment]
@fdwr I've addressed your comments in the new commit, please take another look, thanks!

File: src/lib/validate-input.js [hunk 3]
```diff
@@ -611,6 +611,44 @@ export function validateGatherParams(input, indices, {axis = 0} = {}) {
...
+  }
+  if (!updates.shape.every((value, index) => value === indices.shape[index])) {
+    throw new Error(`Invalid updates value - updates shape should be same as indices shape.`);
+  }
+  if (axis !== undefined) {
+    if (!Number.isInteger(axis) || axis < 0 || axis >= inputRank) {
+      throw new Error(
+          `The axis ${axis} should be an unsigned integer in the interval [0, ${inputRank}).`);
+    }
+  }
```
Comment (huningxin) [inline src/lib/validate-input.js:641]: Should we set default axis value if axis === undefined?

**Review by huningxin — APPROVED** [review]
LGTM with a comment

File: src/scatter_elements.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+  for (let indicesIndex = 0; indicesIndex < sizeOfShape(indices.shape); ++indicesIndex) {
+    // output[indices[i, j, k, ...], j, k, ...] = updates[i, j, k, ...] // if axis == 0
+    // output[i, indices[i, j, k, ...], k, ...] = updates[i, j, k, ...] // if axis == 1
+    // output[i, j, indices[i, j, k, ...], ...] = updates[i, j, k, ...] // if axis == 2
+    const indicesLocation = indices.locationFromIndex(indicesIndex);
+    let indiceValue = indices.getValueByIndex(indicesIndex);
+    indiceValue = indiceValue < 0 ? indiceValue + input.shape[axis] : indiceValue;
+    const outputLocation = indicesLocation.slice();
+    outputLocation[axis] = indiceValue;
```
Comment (huningxin) [inline src/scatter_elements.js:28]: To align with gatherElements, could it be?
```suggestion
    const outputLocation = [...indicesLocation.slice(0, axis), indicesVlaue, ...indicesLocation .slice(axis + 1)];
```

File: src/lib/validate-input.js (hunk 3 above)
Comment (BruceDai) [reply src/lib/validate-input.js:639]: Thanks Ningxin!
I've already set default `axis` as `0` when defining `scatterElements(input, indices, updates, {axis = 0} = {})` function. So  above `if` condition - `if (axis !== undefined)` is always `true`, no need to check that. I removed it in new commit.

**Review by BruceDai — COMMENTED** [review]
Thanks @huningxin!
I've addressed your comments, please take another review in new commit.

**Review by huningxin — APPROVED** [review]
LGTM!

File: src/scatter_elements.js [hunk 5]
```diff
@@ -21,12 +21,12 @@ export function scatterElements(input, indices, updates, {axis = 0} = {}) {
...
-    let indiceValue = indices.getValueByIndex(indicesIndex);
-    indiceValue = indiceValue < 0 ? indiceValue + input.shape[axis] : indiceValue;
-    const outputLoc = indicesLoc.slice();
-    outputLoc[axis] = indiceValue;
-    output.setValueByLocation(outputLoc, updates.getValueByIndex(indicesIndex));
+    const indicesLocation = indices.locationFromIndex(indicesIndex);
+    let indicesValue = indices.getValueByIndex(indicesIndex);
+    indicesValue = indicesValue < 0 ? indicesValue + input.shape[axis] : indicesValue;
+    const outputLocation =
+        [...indicesLocation.slice(0, axis), indicesValue, ...indicesLocation .slice(axis + 1)];
```
Comment (fdwr) [inline src/scatter_elements.js:28]:
```suggestion
        [...indicesLocation.slice(0, axis), indicesValue, ...indicesLocation.slice(axis + 1)];
```
Odd space before ` .slice`.

(personally I found the `outputLocation[axis] = indiceValue` less syntactically noisy and less mental effort to understand and validate, but shrug).

**Review by fdwr — COMMENTED** [review]
👍 Thanks Bruce.

File: src/scatter_elements.js (hunk 5 above)
Comment (BruceDai) [reply src/scatter_elements.js:28]: Good catch! Thanks @fdwr!
Fixed it by #107.

File: src/scatter_elements.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  for (let indicesIndex = 0; indicesIndex < sizeOfShape(indices.shape); ++indicesIndex) {
+    // output[indices[i, j, k, ...], j, k, ...] = updates[i, j, k, ...] // if axis == 0
+    // output[i, indices[i, j, k, ...], k, ...] = updates[i, j, k, ...] // if axis == 1
+    // output[i, j, indices[i, j, k, ...], ...] = updates[i, j, k, ...] // if axis == 2
+    const indicesLocation = indices.locationFromIndex(indicesIndex);
+    let indicesValue = indices.getValueByIndex(indicesIndex);
+    indicesValue = indicesValue < 0 ? indicesValue + input.shape[axis] : indicesValue;
+    const outputLocation =
+        [...indicesLocation.slice(0, axis), indicesValue, ...indicesLocation .slice(axis + 1)];
```
[the remark this answers is not in the document]
Comment (fdwr) [reply src/scatter_elements.js:28]: > BTW, DirectML requires "IndicesTensor.Sizes must match InputTensor.Sizes for every dimension except Axis." https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_scatter_operator_desc

It does? I'm just seeing the requirement that input dimensions == output dimensions, and indices dimensions == updates dimensions.

> Should we restrict indices sizes are less than or equal to input sizes along every dimensions except axis?

Pigeonhole principle dictates there will be overlap of elements, but this only applies to overwrite mode. In general, it should be legal to have indices and updates tensors that are larger than the input (or `data` tensor as CoreML calls it) with other modes like `add`, `sub`, `mul` (note DirectML doesn't support those combination modes).

**Pull request merged**